Side-effect-free Yul interpreter

quangloc99 opened the pull request:
This interpreter implementation is following this specification from #15435 , and is intended to use in #15358 .

This PR has followed spec for the implementation.

For testing, the existing `YulInterpreterTest` was not used. This is because this pure interpreter has no memory/storage access, it can not use the same trace as the current interpreter from 'test/libyul/tools'. But to keep it minimal, the `YulInterpreterTest` test suit was copied and applied minimal changes. The new test suit now uses the function call trace and the outter most variable values for snapshot comparison. Most of the existing test from the `YulInterpreterTest` was ported to the new test suit. New tests were also added to test all return status, as well as to test all evm instructions.

quangloc99 committed:
Copy interpreter tool from test as basis

quangloc99 committed:
Add new Result type

quangloc99 committed:
Change interface of the interpreter to have custom return type

quangloc99 committed:
Implement visit function with new return type

quangloc99 committed:
Remove runExternalCall

quangloc99 committed:
Minor comment edit

quangloc99 committed:
Implement visit(Expression) for ExpressionEvaluator

quangloc99 committed:
Force incrementStep logic to be done at visit function

quangloc99 committed:
Merge ExpressionEvaluator into Interpreter

quangloc99 committed:
Reorder methods

quangloc99 committed:
Reset expressionNestingLevel after each incrementStatementStep

quangloc99 committed:
Remove redundant makeInterpreterNew function

quangloc99 committed:
Remove unused disableMemoryTrace

quangloc99 committed:
Remove state with side effect from InterpreterState

quangloc99 committed:
Move Interpreter Config to a separate struct

quangloc99 committed:
Remove redundant numInstance

quangloc99 committed:
Add recursion depth check

quangloc99 committed:
Remove functions with side-effect from EVMInstructionInterpreter

quangloc99 committed:
Declare ImpureBuiltinEncountered and EVMInstructionInterpretedResult

quangloc99 committed:
Return ImpureBuiltinEncountered for all builtin with side-effect

quangloc99 committed:
Use EVMInstructionInterpretedResult in Interpreter

quangloc99 committed:
Bring back empty body and post check when visiting ForLoop

quangloc99 committed:
Remove unused function in EVMInstructionInterpreter

quangloc99 committed:
Move result types into a separate file

quangloc99 committed:
Move PureInterpreterState into a separate file

quangloc99 committed:
Add Pure* prefix

quangloc99 committed:
Add Results and PureInterpreterState into CMakeLists

quangloc99 committed:
Separate function definition and variable declaration in Scope

quangloc99 committed:
Make Scope parent constant pointer

quangloc99 committed:
Move Scope to a separate file

quangloc99 committed:
Make Scope class. Add utilities to Scope.

quangloc99 committed:
Use new Scope class in PureInterpreter

quangloc99 committed:
Create VariableValuesMap alias

quangloc99 committed:
Fix behavior of POP instruction

quangloc99 committed:
Add stack trace to state

quangloc99 committed:
Remove TraceLimitReached from ExecutionTerminated

quangloc99 committed:
Add dumpTraces function

quangloc99 committed:
Add execute statements function

quangloc99 committed:
Add UnlimitedLiteralEncountered result type

quangloc99 committed:
Change calculation flow for evaluateArgs

quangloc99 committed:
Add move constructor for EvaluationOk

quangloc99 committed:
Disable EvaluationOk lvalue constructor

quangloc99 committed:
Make EvaluationOk constructor explicit

quangloc99 committed:
Optimize std::vector<u256> usage

quangloc99 committed:
Copy YulInterpreterTest suite to YulPureInterpreterTest

quangloc99 committed:
Read interpreter config from test settings

quangloc99 committed:
Make PureInterpreter.run returns ExecutionResult

quangloc99 committed:
Add default value for traces

quangloc99 committed:
Use new PureInterpreter in test

quangloc99 committed:
Make TraceLimitReached a ExecutionTerminated result again

quangloc99 committed:
Remove PureInterpreter.run method

quangloc99 committed:
Add function to get allVariables from interpreter

quangloc99 committed:
Dump more data after interpreter

quangloc99 committed:
Register YulPureInterpreterTest

quangloc99 committed:
Filter and add existing interpreter tests to pure interpreterTest

quangloc99 committed:
Add hex printing and sort variable names

quangloc99 committed:
Add tests for all impure instructions

quangloc99 committed:
Add test for all pure instructions

quangloc99 committed:
Change expr_nesting_depth_not_exceeded to run expression multiple time

quangloc99 committed:
Add test for StepLimitReached

quangloc99 committed:
Add test for impure non instructions

quangloc99 committed:
Add check for verbatim

quangloc99 committed:
Add verbatim test

quangloc99 committed:
Remove redundant public:

quangloc99 committed:
Optimize map usage

quangloc99 committed:
Remove redundant terminated check

quangloc99 committed:
Remove redundant terminated result

quangloc99 committed:
Add test for stop instruction

quangloc99 committed:
Update tests to remove trailing space

quangloc99 committed:
Fix styling for addTrace

quangloc99 committed:
Fix spelling

quangloc99 committed:
Fix linting error for generate-test.py

quangloc99 committed:
Update EVMVersion in tests

quangloc99 committed:
Add encoding when open file when generate test

quangloc99 committed:
Fix spelling Expection -> Expectation

quangloc99 committed:
Remove virtual modifier from PureInterpreter functions

quangloc99 committed:
Remove unused m_state

cameel commented on libyul/tools/interpreter/PureEVMInstructionInterpreter.cpp:
```diff
@@ -0,0 +1,312 @@
+	if (isVerbatim)
+		return ImpureBuiltinEncountered();
+
+	static std::set<std::string> const NON_INSTRUCTION_BUILTIN_NAME = {
+		"datasize",
+		"dataoffset",
+		"datacopy",
+		"memoryguard",
+		"loadimmutable",
+		"setimmutable",
+		"linkersymbol"
+	};
```
Instead of hard-coding those, you should just check that the `instruction` field in `BuiltinFunctionForEVM` is empty.

quangloc99 replied:
I will change as suggested. I was just afraid that there might be builtin other than instructions that can be evaluated in the future. This part was added to have the same runtime warning effect as I was doing with the switch exhaustive check.

cameel replied:
If there is, we'll quickly find out in tests and fix it, so don't worry about it.

The only other thing are IIRC the dynamically generated verbatim functions, but you reject those as well.

quangloc99 replied:
This got fixed, along side making the `PureEVMInstructionInterpreter` a function.

cameel commented on test/libyul/YulPureInterpreterTest.cpp:
```diff
@@ -0,0 +1,188 @@
+	m_source = m_reader.source();
+	m_expectation = m_reader.simpleExpectations();
+
+	m_config.maxTraceSize = m_reader.sizetSetting("maxTraceSize", 128);
+	m_config.maxExprNesting = m_reader.sizetSetting("maxExprNesting", 64);
+	m_config.maxSteps = m_reader.sizetSetting("maxSteps", 512);
+	m_config.maxRecursionDepth = m_reader.sizetSetting("maxRecursionDepth", 64);
+
+	m_printHex = m_reader.boolSetting("printHex", false);
```
> For testing, the existing `YulInterpreterTest` was not used. This is because this pure interpreter has no memory/storage access, it can not use the same trace as the current interpreter from 'test/libyul/tools'.

That's actually easy to solve. Just add a setting you can set to mark a test as one that requires access to memory or storage. Then add it to those tests that do not pass without it. As you can see in the snippet above, adding settings is pretty easy.

For this we could call the setting `sideEffectFree` and it `true` by default.

quangloc99 replied:
I understand how to mark a test that requires memory access. However, I’d still insist to keep the two test suites separate. While they do share the parsing part, the settings and trace printing are quite different. I’m concerned that merging the two might make it harder to manage the tests.

I also believe the two interpreter implementations might grow in different directions, so separating the test suites makes more sense to me.

cameel replied:
While we're ok with leaving the interpreters separate for now to avoid bringing over unnecessary baggage from the test interpreter, the test suite should definitely be shared. Even if the implementations diverge significantly, they ultimately both use Yul as their interface so it should not matter for testing. The same set of tests should be usable for both and I strongly think this is the point we should be starting from. Seeing the new interpreter pass the existing tests (at least the ones it can run) will be an additional confirmation that it works correctly.

I also don't see why settings and trace would have to be substantially different. If the new ones are better, we should just change them in the old test suite as well. I'd suggest actually extracting that into a PR of its own, which we'd merge first.

quangloc99 replied:
I agree that the testing interface is the same. I will adjust the test suit to use the existing one.

The settings is also not that different. The only notable settings that the pure interpreter has is the _max recursion depth_. I decided to add this, because this interpreter will be used while walking the AST. This allows the code to be interpreted with a big `maxStep`, while stack memory is still protected from overflowing.

The trace is definitely different. The trace of the interpreter in `test/tools` consists the following:
- Storage layout
- Memory layout
- Transient layout
- History traces about memory access, log

The pure interpreter does __not__ evaluate instructions that alter memory/storage, or produces log. Therefore such trace can not be created. I have changed the trace in to the call trace and the values of the outer most variables.

The new trace can be included into the `test/tools` interpreter, but not the other way around.

quangloc99 replied:
I have another suggestion. I will merge the two testsuit together. For each test, I will print the traces from both interpreter.
How does this sound?

cameel replied:
> The only notable settings that the pure interpreter has is the *max recursion depth*.

Do we need it as a test setting though? I'm assuming that this will be a hard-coded limit in the compiler, so why use a different one in tests?

Or do you intend to expose it to the user as a setting? We generally rarely expose such settings and I don't think this one would be relevant enough to warrant it. After all, this will never even occur in the main use case, which is `constant`s.

> The pure interpreter does **not** evaluate instructions that alter memory/storage, or produces log. Therefore such trace can not be created.

We can make the pure interpreter always print those the same way the current interpreter does when it's running on pure code. Or we could make the existing interpreter not show this when the traces are empty. This does not seem like a significant obstacle to unifying them.

> I have another suggestion. I will merge the two testsuit together. For each test, I will print the traces from both interpreter.
> How does this sound?

But one will always be empty so why even print it? I think that having the existing interpreter only print the non-empty trace would be the better way to handle this.

quangloc99 replied:
> I'm assuming that this will be a hard-coded limit in the compiler, so why use a different one in tests?

I forgot that we indeed hardcode the limit in the compiler. In the test, I tried to make it as flexible as possible. But since we'll hardcode anyway, I'll remove settings and test with a known limit.

> We can make the pure interpreter always print those the same way the current interpreter does when it's running on pure code. 

I still dont see the point of this. The trace of the current interpreter is all about state changes. The pure interpreter does not touch the state, so there is no such trace to print.

> Or we could make the existing interpreter not show this when the traces are empty.

In case of empty trace, the current interpreter does have zero trace about state change. But by default, the pure interpreter also have no trace about state change as well. That's why I added another kind of trace so we can see if it works correctly.

>> For each test, I will print the traces from both interpreter.
>
> But one will always be empty so why even print it?

Not exactly.
- For the case the code has state change, the pure interpreter will still have the call trace up to the point where the state change happens, and the `ImpureInstructionEncountered` will be reported.
- For the case where there is no state change, the current interpreter will not have any trace. But we know that for this test, the current interpreter will still do its job.

My idea for the suggestion to print both traces is to test both interpreter at the same test. Both interpreter can run the same yul code, just with different terminated status and traces. And I think this work wrll for the PR: it will use the same testsuit, but it does not need to touch the code for the current interpreter.

cameel replied:
Ah, so it looks like your trace and the trace in the current interpreter are different things. I assumed they were the same and that your trace was just a subset of the old one. Looks like instead you're recording function calls, while the current interpreter records builtins?

Another thing is that I assumed that the trace was a feature of the test case, not the interpreter itself. It's a debug feature, and this made sense in an interpreter built for testing, but for executing *correct* code in the optimizer that's pure overhead. I think that this should be done by having the base interpreter provide virtual functions as hooks, and having a derived test-only variant that stores the trace.

Overall, there's no technical reason why both interpreters couldn't provide the same output and IMO the best solution would be to just extend the current interpreter to report function calls too. But ok, that's extra work, which I didn't expect this would require so I guess just printing one kind in one interpreter and the other kind in the other is acceptable for this PR. We should give them different names though, because they're really different things.

quangloc99 replied:
Thank you for understanding!

> ... so I guess just printing one kind in one interpreter and the other kind in the other is acceptable for this PR

With this, I'll merge the 2 test-suits into one, and print both traces for the same test case.

> We should give them different names though, because they're really different things.

I am not sure whose _name_ are you referring to. You mean the 2 interpreters should have different names, or their traces should have different names?

> ... IMO the best solution would be to just extend the current interpreter to report function calls too

That what I have originally done in #15358 :pray: . I moved the interpreter from `test/tools` into `libyul` and have the pure interpreter extending that interpreter. I did not change the trace though but that can also be done.

> ... but for executing correct code in the optimizer that's pure overhead

I am positive that my way of printing trace do not have overhead. Here is [godbolt the demo for my trace][trace demo]. If we hardcode `maxTraceSize` to 0, compiler should eliminate all `addTrace` calls.

[trace demo]: https://godbolt.org/#z:OYLghAFBqd5QCxAYwPYBMCmBRdBLAF1QCcAaPECAMzwBtMA7AQwFtMQByARg9KtQYEAysib0QXACx8BBAKoBnTAAUAHpwAMvAFYTStJg1DIApACYAQuYukl9ZATwDKjdAGFUtAK4sGIMwAcpK4AMngMmAByPgBGmMQgAdIADqgKhE4MHt6%2B/kGp6Y4CYRHRLHEJSbaY9kUMQgRMxATZPn6B1bWZDU0EJVGx8YnSCo3Nrbkdo739ZRXDAJS2qF7EyOwc5gDM4cjeWADUJltuAG6YDiTH2CYaAILbu/uYRydOo8SYrNe3D2Y7DD2XkOxzOTTwhgIP3uj0Bz1eHmSdTE0Ie9wA9OiDvxiAcsCwBB8mHUDgJaABPX5edJGA4AMS8gLqABFMDQGBkBK9mQdRugQCAPuFgMcrPdqcKDl4zABWABs3IOtAEwCVKtFv019z2TAUCgOABViEx1mEWIQAEpfZAITDoTUAdjFDwdzI1MI9dw%2BXgc9MZDkybjEtCNJswjudtwAnAymZlWezOQwDmgGKNzAqsIm6u60VG%2BQLzpdiKDpfLrgdkk1WApc1qY/66kHaCHjesIPXo7GAwIE%2BEkynCVCzAqAPpZ/t1UidjT5gj8kBFoglk5luUV1PpkcHUdV40sWseqMLEAzqMTjl1CDjtmTzJLM97mvXp8Hhadp0mV2Ot1bSP3b1fW7OorQIVYGFDdYI3rYD41vS9MkHNNh0zeCkzrWcC0XC5l1LWV1y2bADk%2BMDiAidAADUxC8TBaz/LU81ggRQPAyDMA7I8u0bODs0QzcUJ3C8k2nTjMPnQscKuVd8I3IcMx3EjwLtKjvFo%2BsTzPISrxvXiBAfUSo0UsjlOo2jryM8iVJohR304z9vxhb96M9CVaRCVBgDY7BBGIclFSw05wUhUEmKyYM2NIP042YzBSIgttw0IjCvQIYgfQIA5lFWTAAElBHiZISPiDwGBoEVHP/KNMV5WKDiIA4NDq1A8TwBQmBieg6rbYV63SAAvTBRwylgmFUNihDwAbFQ0ZLo36wbhtGhpMGSfVjh5GbnLzeahoOEbVGwVRCsiWjHFpdaGow/NJoWvbRqtZBVnSXsVoIBBprrJz/zRFK0t9LLPjyggCqK4gemB6CjwB3L8uIQrYuKgQyqQ0YkLK2bMJu3aGB8ZbVo%2Bra5vE7Di1BdzPIS7zUspQiurDNath5L8rC%2Bhjo0xLE7nQdADiYNVVVSsMADp6w5g4AHU6FoPFmoYVA3slPAqDRvBgCF/bxpu7kLpmo8OejABaOrbX5g5GGpg4AHcpYOOWMriFHUvSu0DnCY3WpTXVwyPYGWGSAwIZOAhyWSRhWBeNiDRDzBIuD0PmDYIWk4OO5iGAOibiPLDUCRTIURONizUta1bXtWmmG5tiIFT9PHYzJOhZ3Jp09svNmbPJWDggVMyvV0bNam9adYWYjYvA3liZxlsc6hQnZ2jTuIEF9YFCF%2BaIBH2TStVvuxoSiaBtb6Nj9nCzDQSouCCtE1S43q7o2X2ihcwP2DHWUcYhNABrCAsPCXc34LWjqCSOwDCKRVHM3VeSd3xzyjNGM%2BWEp7KiRPfByP1oynFQHgHm6AfDJDYgoX%2BxM0ipS%2BCweSo4VgEBHvxT6v5vrhEWuEDeRxPz3AOJwqUNJVQJ1olWdYE8y7fS4ZlbKQMQYI2ICVZGPdVa5lEXItWGt95ax1qKTh1VQQGx0bovReiUwIEMMAF4Ct9T1SUKYk2bIqA4QYqI6GEi4ag3Bi8aYEMnRKK/Aw%2BxXChS0ioMmC65gzD8FQCEhRXClxSTcGuCsr4GZMydFwSKZhIpbG8ZEzh0SVyxJkrTCyJlVKJLYRYaQBwZSRXXKzDhfjGjAyFhXdAbEQrcQEM2VsYZrjUAYJFBJsDnSiPcZgRplcEqtOigwFiZEWmER6ZFQplFTI2WSqInEXcmBeHqnQ7cj8QAT2JCMx%2BNlSm/FEaIxemyiAACpsRhRbIqExBBRxKwmT2e5nSoJzIzI/BYm92F3HOUCwcWyESgjuR0g21wtKIVBOCkJnsWwu2tm9I4ZgzBgpOBC4MULCIJLXjdVhcKsUItfFI1eCLiVuDNgwdAtAsnnO8WbWgShXbKwgFc1AtzAnTKCYzA4TyXlUDeSBMeMzxnfJHL8/5gzgWKOoZi6lPKxUMFxbgNCJIqVooxYsq2hB3qUpOOC5VcU1WLKsk/dem8jUkvRaPOKLtArFJFnarVrh6VbTlZk2pnDvUumZBwJYtBOAyl4H4bgvBUCcDcNYawvIVhrBeNsHgpACCaEDUsL%2BIAZQyiFoEAIUYAgygCP8LYWw5RRgdPoTgkheAsAkBoDQpBw1aFIFGjgvAFAgCbWmjgWglhwFgDARAKBUB%2BzoPEcglA0DjvoAkYAUg0k0FoMDYgXaIAxHTaQGI4Qmjkk4CmndzBfIAHkYjaBwge3gM62CCBPQwCkW6sAxC8MADpXaI2kHxMY8QfbeD4E%2BAGc4H7W2YFUBcLZGwU1MJqFe/QeAYjGl8h4LAcHUp4HrX%2B0g5xiAxDSJgVkI0jC0GFOmpYVADDpwongTAlsT3xzg/wQQIgxDsCkDIQQigVDqCw7oFJBgjAoFjZYeDMQu2QCWDPTIH6DZ8nWqYSw1guAOgOAbE9ZhI04eIDg2i8Alh2BwpkFwtKJh%2BBSaEcIAxyhDBSQUJMpm9B2bqLMQYCQUkGfeT0cYng2h6A83ULzfRLNzBs7YMYLQfO5Hc%2BFlz1m3P6YTesCQQaQ1hq3e2g4qgAhygNnKSQArkDIAOFIPNXdcCEBIGirYXAFi8F7f2pYtoK5DA3qQLNZgox5qjJIDQkgzAOhlBoOUcouBcDlFsatHBa2kHrQ6B0QsNBcF60NgIo2HTVeq829LnBO3dtTWR0gg6R3UOSFsqd3cx3JAncQSI4cQAhEOaMYI%2BBlw4L0Ex4QohxDsY%2B1xtQW6%2BOkEtsaZIV6UscFDVtrD7aT1bNOxlVAysss5bywVorJWMUQA8LO%2BIVWat1bI41r4WAEiteDVNut/gzBC0LVGGUUZqsBBLXKbLRaoetvbbtnthO2sgEkJIIWDpRuLckMp4b43euTa2Gl6HO39t/r%2BZNjTM2G1NpbZGuX9WM3YfiM9PwkggA%3D%3D%3D

cameel replied:
> I am not sure whose *name* are you referring to. You mean the 2 interpreters should have different names, or their traces should have different names?

Traces. Just make it clear that one shows the function call stack while the other one is the record of evaluated builtins.

> That what I have originally done in #15358 🙏.

Yeah, and that still would be an option if you want to do it. But I don't want to add too much extra work here. Just printing both things is fine for now too.

> If we hardcode `maxTraceSize` to 0, compiler should eliminate all `addTrace` calls.

That's good, bit I'd still rather have the trace functionality only in the test case. The less complexity in the actual implementation, the better.

cameel replied:
Anyway, my suggestion would be to submit your test cases and the changes in the `YulInterpreterTest` as a separate PR so that we can review and merge that first. They should already all pass on the existing interpreter. Then we can continue with this PR.

cameel commented on libyul/tools/interpreter/PureInterpreter.cpp:
```diff
@@ -0,0 +1,360 @@
+		// Increment step for each loop iteration for loops with
+		// an empty body and post blocks to prevent a deadlock.
+		if (_forLoop.body.statements.size() == 0 && _forLoop.post.statements.size() == 0)
+			if (auto terminated = incrementStatementStep()) return *terminated;
```
Some general notes on style to get that out of the way:
- We always put `return` on a new line.
- Please also avoid abbreviations in names. Things like `fun`, `cnt`, `x`, `g`, `vec`, `res` are hard to read and often ambiguous. TBH, we have a lot of bad names already since the code is old and it's also hard to enforce, so some of that might have been just copied, but when writing new code, please try to avoid them.
- Constructor spacing: `ExecutionOk{ControlFlowState::Default};`
- 4-space indents, including in `.yul` files.
- You don't have to include the empty string as message in asserts. We made it optional at some point, but did not strip it from everywhere. When writing new code, you can omit the message.
- I'd avoid making files hidden (with a leading dot in the name).
- In Python code we use the same style for splitting long calls as in C++. I.e. 
    ```python
        gen_test(
            'sgt',
            param_cnt=2,
            calc=lambda p: u2s(p[0]) > u2s(p[1])
        )
    ```

quangloc99 replied:
Thank you for commenting on the styling. For this part, I have a question about the formatter.

For C++ code style, I see that there is `.clang-format` configuration included. However, using clang-format changes the code drastically. So I deliberately turned off the formatter. Should I use clang format for the new code, or should I try to adjust the code manually? And also the same question applied to Python code.

The is the only questions. I will address the other comments.

quangloc99 replied:
Besides the code style, is it OK to use macro for the _terminated_ case? I think that check is repetitive, and because it also `return` immediately, I can only think of macro to clean up the code.

cameel replied:
> Should I use clang format for the new code, or should I try to adjust the code manually?

No, that's an incomplete config that does not fully reflect our style unfortunately. There is an effort underway to get to a workable clang config, but for the time being we format the code manually.

> And also the same question applied to Python code.

For Python we also generally format manually, but our style is generally very close to the one used by [black](https://github.com/psf/black), so if you use that, it should be fine.

quangloc99 replied:
I have resolved most of the comments for the styling. Hopefully I did not miss any renaming.

I still have the pending question on using macro.

And also I'm not sure where to put the test generator for the `pure` instructions. My idea was that I want to test the calculation of those instructions across some test, against a different implementation (in this case is my Python test generator). I am not sure if there a better way to do this. For now it seems to me that putting the test generator near the tests is still the best choice. Please give some advice for this.

cameel replied:
> I still have the pending question on using macro.

Sorry, I saw it, but forgot to answer before moving on to the next comment.

I'm not sure about that one yet. I'm not a fan of hiding control flow like this and it seems like there aren't *that* many places where you check this, so I don't see it as a big problem. It seems quite manageable, especially since the interpreter is the only place where this style is used. Maybe a helper to simplify the check bit would be enough?

```c++
EvaluationResult result = evaluate(_expressionStatement.expression, 0);
if (terminated(result))
	return result;
```

On the other hand, the old interpreter used exceptions and that also hides the flow, so a macro may not be worse than that. I do agree that these checks seem tedious and easy to forget so I wouldn't completely reject the idea. An ideal solution would be something like `?` in Rust, but we don't really have an equivalent in C++. For now, the helper approach seems to me like it might be a good compromise.

> And also I'm not sure where to put the test generator for the `pure` instructions.

Not sure about this one either. We don't have a standard practice for this. We generally put scripts into the `scripts/` directory, but then that would break the connection between the generator and those tests. On the other hand, even if it's alongside the tests, I doubt we'll generally pay attention to it in the long term anyway. That's just what happens with these things - we'll likely end up with manual modifications in these tests over time and you won't be able to safely regenerate them without losing those modifications. It may be best to treat the script more as documentation on how they were initially generated rather than something that we'll commit to maintaining in the repo. As such it might be better to just post the script in a comment in the PR.

quangloc99 replied:
For the macro part, I'll use macro.

> An ideal solution would be something like ? in Rust, but we don't really have an equivalent in C++.

Well this is one reason why I proposed using boost's `outcome` instead of std::variant. They have well documented [TRY macro](https://www.boost.org/doc/libs/1_75_0/libs/outcome/doc/html/tutorial/essential/result/try.html) that looks almost natural as Rust `?`. If you like this idea then I guess I can use it too.

For the python test generator, I'll put it into the comment for the corresponding test it generated.

cameel replied:
> Well this is one reason why I proposed using boost's `outcome` instead of std::variant. They have well documented [TRY macro](https://www.boost.org/doc/libs/1_75_0/libs/outcome/doc/html/tutorial/essential/result/try.html) that looks almost natural as Rust `?`. If you like this idea then I guess I can use it too.

Let's try that then. I actually do like the idea of trying something from a well established library for this rather than hand-implementing it. It has a better chance of being recognized as something conventional and not being as surprising despite the unusual control flow.

> For the python test generator, I'll put it into the comment for the corresponding test it generated.

Regarding that, one alternative solution that came to my mind is that we could have one of the CI jobs (e.g. the one that tests Python scripts, since that's guaranteed to have Python) run the script and fail if changes are detected. With this it would actually be acceptable to have it in the repo, because we could rely on people not editing these tests randomly.

I'd also make the script insert a comment at the beginning of the test, saying that it was automatically generated and how to rerun it.

quangloc99 replied:
I have changed to boost's outcome. Please take a look.

cameel reviewed (commented):
Thanks for the PR! I won't manage to do a proper review until next week, but for now here's some initial feedback from a quick pass over it.

cameel commented on libyul/tools/interpreter/PureEVMInstructionInterpreter.cpp:
```diff
@@ -0,0 +1,312 @@
+	case Instruction::SWAP13:
+	case Instruction::SWAP14:
+	case Instruction::SWAP15:
+	case Instruction::SWAP16:
+	{
+		yulAssert(false, "");
+		return EvaluationOk(0);
+	}
+	}
+
+	yulAssert(false, "Unknown instruction with opcode " + std::to_string(static_cast<uint8_t>(_instruction)));
+	return EvaluationOk(0);
```
```suggestion
	case Instruction::SWAP16:
		yulAssert(false, "Instruction not allowed in strict assembly.");
	}

	util::unreachable();
```

quangloc99 replied:
Thank you for pointing out the `util::unreachable`. I actually don't know about it.

However, I think the assert with error message should be used. Currently all the instruction are checked with a `switch` statement, which is not guaranteed to be exhaustive check at compile time. I am not sure if there is a way to ensure its correctness at compile time. For the runtime check, I think it should fail loudly. When failure happens, there must be a new instruction. It should then be added to this function.

cameel replied:
> which is not guaranteed to be exhaustive check at compile time

If you do not include a `default` case, most compilers (at least GCC and Clang) will warn when you don't handle all cases. We use warnings-as-errors by default so it will just not compile instead of failing at runtime. So correctness in fact *is* ensured here at compilation time already.

Actually, the compiler should even figure out on its own that any code after the `switch` is unreachable, and usually it does. `unreachable()` is just our workaround for a GCC bug (see #13087) that makes it fail to do so in Debug mode.

quangloc99 replied:
I changed to `util::unreachable()`. Thank you for the insight about the workaround! :pray:

cameel commented on libyul/tools/interpreter/PureEVMInstructionInterpreter.cpp:
```diff
@@ -0,0 +1,312 @@
+
+EvaluationResult PureEVMInstructionInterpreter::evalBuiltin(
+	BuiltinFunctionForEVM const& _fun,
+	std::vector<Expression> const& /* _arguments */,  // This was required to execute some builtin.
+													  // But all of them are impure.
```
Are they though? I'd expect `dataoffset()`, `datasize()` and `linkersymbol()` at least to be pure. We won't be able to evaluate them without information that's only available at bytecode generation time, but technically they end up being just constants :)

quangloc99 replied:
I understand the point. I think my comment here is misleading too, so I will change that.

I think for the implementation, we can try returning special values for these cases. But that sounds more complicated, and it is go beyond the point of code interpretation.

cameel replied:
It's probably fine not to do anything about it. I don't think we'll be able to actually implement those. Just wanted to correct the comment :)

quangloc99 replied:
I removed the comment about _impure_.

cameel replied:
You can actually remove the whole argument. It was used for literal arguments of `datasize()` and `dataoffset()`, which the new interpreter does not support.

Literal arguments are basically compile-time-only values. They're not physically stored on the stack and may not even fit in a slot. This means we can't pass them around as `u256`. When the value is needed, it is obtained from the relevant AST node instead. Here it will never happen though.

quangloc99 replied:
This is removed since I make `PureEVMInstructionInterpreter` a function.

cameel commented on libyul/tools/interpreter/PureEVMInstructionInterpreter.cpp:
I'd drop the `tools/` dir and put it simply under `interpreter/`. The old interpreter is under `test/tools/` because it's a part of `yulrun`, which is a testing tool (and compiles into a separate executable). The new interpreter becomes a part of `libyul` instead, which goes into `solc`.

quangloc99 replied:
I have dropped the `tools` dir.

cameel commented on libyul/tools/interpreter/PureEVMInstructionInterpreter.cpp:
```diff
@@ -0,0 +1,312 @@
+using solidity::util::h160;
+using solidity::util::h256;
+
+using u512 = boost::multiprecision::number<boost::multiprecision::cpp_int_backend<512, 256, boost::multiprecision::unsigned_magnitude, boost::multiprecision::unchecked, void>>;
```
Why not place this next to the other types like this we already have, like `u256`?

quangloc99 replied:
I did not notice that type was there. I was too focused on modifying the original interpreter :pray: .

I will move it.

quangloc99 replied:
I have move `u512` to `libsolutil/numeric`. However, I have found a minor thing with the type definition: the `MaxBits` is `256`, while `MinBits` is 512. This is fixed in the commit https://github.com/ethereum/solidity/pull/15464/commits/5351aa8905dc12b4b1a603236adb2b78a59bf8ac .

As I was copy and pasting the implementation before adding my change, the original interpreter in the `test/` dir has the same _issue_.

Currently the _issue_ does not lead to any error. I checked boost implementation, and the precision is obtained by getting the max of `MaxBits` and `MinBits` (if the allocator is `void`). I think maybe we can just use the predefined types from [boost cpp_int](https://www.boost.org/doc/libs/1_84_0/libs/multiprecision/doc/html/boost_multiprecision/tut/ints/cpp_int.html) instead of redefined like this.

cameel replied:
> I have move `u512` to `libsolutil/numeric`. However, I have found a minor thing with the type definition: the `MaxBits` is `256`, while `MinBits` is 512. This is fixed in the commit https://github.com/ethereum/solidity/pull/15464/commits/5351aa8905dc12b4b1a603236adb2b78a59bf8ac .

Nice. Can you submit it as a separate PR (along with moving `u512` to `libsolutil`)? We can merge it as a bugfix.

> I think maybe we can just use the predefined types from [boost cpp_int](https://www.boost.org/doc/libs/1_84_0/libs/multiprecision/doc/html/boost_multiprecision/tut/ints/cpp_int.html) instead of redefined like this.

Sure, why not. I'm not sure why this wasn't done in the first place. Maybe these typedefs did not exist back when we defined these types?

You can submit this as a PR too.

quangloc99 replied:
I have made the PR here #15510

quangloc99 committed:
Drop tools/ subdirectory

quangloc99 committed:
Move u512 to libsolutil/Numeric.h

quangloc99 committed:
Remove comment about instruction being _impure_

quangloc99 committed:
Use util::unreachable in PureEVMInstructionInterpreter::eval

quangloc99 committed:
Fix styling: return on separate line

quangloc99 committed:
Fix styling: spacing surround constructor call

quangloc99 committed:
Fix styling: format python code

quangloc99 committed:
Fix styling: yul test indentation

quangloc99 committed:
Remove empty string message in assert

quangloc99 committed:
Make variable name explicit

cameel commented on libyul/interpreter/PureInterpreter.cpp:
```diff
@@ -0,0 +1,378 @@
+	for (size_t i = 0; i < functionDefinition.returnVariables.size(); ++i)
+		variables[functionDefinition.returnVariables.at(i).name] = 0;
+
+	std::unique_ptr<PureInterpreter> interpreter = makeInterpreterCopy(std::move(variables));
```
This looks like it lives only until the end of the function. Why even allocate it on the heap?

quangloc99 replied:
The heap allocation was copied from the original interpreter. Originally I did not think much about it, but then I think making heap allocation should help save more stack space, so we can allow more recursion calls.

But I also added the config to limit the number of recursion calls, so if you think this part does not have much effect on the stack, then I can change it to allocation on stack.

cameel replied:
Yeah, that was my impression too, but then if stack space was a real concern then it's only a half-measure and a proper solution would be to use a non-recursive implementation instead. I think we should just put it on stack and revisit if it turns out to be an actual problem. We should have a test that uses nested calls up to the limit to verify that.

In practice we won't have much nesting anyway, so we could live even with a very low limit.

You'll probably find more questionable decisions like this in the old implementation. If it does not seem justified, we want it simplified. The general idea here is to treat it as a new implementation that's carefully reviewed from scratch.

quangloc99 replied:
This is fixed.

quangloc99 committed:
Move pure_instructions test generator into test comment

cameel commented on libyul/interpreter/PureEVMInstructionInterpreter.cpp:
```diff
@@ -0,0 +1,307 @@
+)
+{
+	using namespace solidity::evmasm;
+	using evmasm::Instruction;
```
These don't seem to be necessary. Could be removed both from there and from the original interpreter.

cameel replied:
The forward declaration of `YulString` in the header seems unused as well. Overall I have an impression that there might be some unused includes and forward declarations left over from the original implementation.

quangloc99 replied:
I have remove unused namespace and includes.

cameel commented on libyul/interpreter/PureEVMInstructionInterpreter.h:
```diff
@@ -0,0 +1,76 @@
+/**
+ * Interprets EVM instructions based on the current state without side-effect.
+ */
+class PureEVMInstructionInterpreter
```
The pure interpreter does not seem to carry much state. How about simplifying it by removing the class and turning `eval()` and `evalBuiltin()` into standalone functions?

quangloc99 replied:
Yes I can do that. My impression when porting this to pure interpreter is because of the dialect. A new instruction interpreter need to be provided if there is a new dialect.

Though I think simple functions will serve the job just fine. I can also scope them inside a namespace too.

quangloc99 replied:
Changed this to a function.

cameel commented on libyul/interpreter/PureInterpreter.cpp:
```diff
@@ -0,0 +1,378 @@
+{
+	std::vector<std::optional<LiteralKind>> const* literalArguments = nullptr;
+	if (BuiltinFunction const* builtin = m_dialect.builtin(_functionCall.functionName.name))
+		if (!builtin->literalArguments.empty())
```
Since we don't support any builtins with literal arguments anyway, you can simplify this by returning `UnlimitedLiteralEncountered` already from here when `!builtin->literalArguments.empty()` and removing `_literalArguments` argument from `evaluateArgs()`.

quangloc99 replied:
`_literalArguments` is removed.

cameel commented on libyul/interpreter/Scope.h:
```diff
@@ -0,0 +1,68 @@
+	std::map<YulName, FunctionDefinition const&> m_definedFunctions;
+	std::vector<YulName> m_declaredVariables;
+	std::map<Block const*, std::unique_ptr<Scope>> m_subScopes;
+	Scope* const m_parent;
```
Our convention now is to always initialize variables of POD types in declaration (this might not have been the case when the original interpreter was written).

We do that even if they get a value assigned later in constructor - in that case I'd just use the default initialization with `{}` to emphasize that the value used here does not matter, only that it does get some deterministic value.

```suggestion
	Scope* const m_parent{};
```

You could use the same style with `PureInterpreter::m_recursionDepth`.

quangloc99 replied:
I have put default initialization to fields as advised.

cameel commented on libyul/interpreter/Scope.h:
```diff
@@ -0,0 +1,68 @@
+{
+public:
+	Scope(Scope* const _parent = nullptr):
+		m_definedFunctions(),
+		m_declaredVariables(),
+		m_subScopes(),
+		m_parent(_parent)
```
```suggestion
		m_parent(_parent)
```

cameel commented on libyul/interpreter/Scope.cpp:
```diff
@@ -0,0 +1,67 @@
+	if (!isNew)
+		return it->second.get();
+
+	it->second = std::make_unique<Scope>(this);
+	Scope* subscope = it->second.get();
+
+	for (auto const& statement: _block.statements)
+		if (auto const* functionDefinition = std::get_if<FunctionDefinition>(&statement))
+			subscope->m_definedFunctions.emplace(functionDefinition->name, *functionDefinition);
+
+	return subscope;
```
This initialization looks like something that belongs in the constructor. Here I'd do just this:
```suggestion
	auto& [block, subScope] = it->second;
	subScope = std::make_unique<Scope>(this, block);
	return subScope.get();
```
and let the scope find its own functions.

Also, I did not expect this function to create a scope since it's called `getSubscope()`. `createSubscope()` would be a more self-explanatory name.

quangloc99 replied:
I have moved the function assignment to constructor, and renamed the function as well.

cameel reviewed (commented):
Not a full review yet, but here's a few things I found looking through the implmentation today.

Mostly suggestions about things that could be simplified or removed.

cameel commented:
The PR needs a rebase.

Also note that some EOF-related PRs were merged recently. There were some minor changes in the iterpreter, please take a look at history. Especially with #15467 you'll probably need to pass `EOFVersion` to the `Dialect`.

quangloc99 committed:
Add DATALOADN to PureEVMInstructionInterpreter

quangloc99 committed:
Adapt to new EOF changes

quangloc99 committed:
Use boost::outcome

quangloc99 committed:
Change makeInterpreterCopy to allocate on stack

quangloc99 committed:
Replace hardcoded builtin with ImpureBuiltinEncountered error

quangloc99 committed:
Make PureEVMInstructionInterpreter just a function

quangloc99 committed:
Return UnlimitedLiteralEncountered early

quangloc99 committed:
Default initialization for fields

quangloc99 committed:
Move function initialization part of scope to scope's constructor

quangloc99 committed:
Rename getSubscope -> createSubscope

quangloc99 committed:
Initialize test subscope with block in test suite

quangloc99 committed:
Remove operator== of Terminated outcome

quangloc99 committed:
Remove unused imports

quangloc99 commented:
I have rebased this branch to the recent develop, and I think I have addressed most of the change request. Please help take a look.

There is still the request about the test suite. I'll add some tests to a new branch, and merge the test suite later.
